aidial_adapter_openai/gpt4_multi_modal/transformation.py

```diff
@@ -1,100 +1,155 @@
-from typing import List, Optional, cast
+from dataclasses import dataclass
+from typing import List, Set, cast
 
 from aidial_sdk.exceptions import HTTPException as DialException
 from aidial_sdk.exceptions import InvalidRequestError
-from pydantic import BaseModel
+from pydantic import BaseModel, Field
 
-from aidial_adapter_openai.gpt4_multi_modal.attachment import (
-    ImageFail,
-    download_image,
+from aidial_adapter_openai.dial_api.resource import (
+    AttachmentResource,
+    DialResource,
+    URLResource,
+    ValidationError,
 )
-from aidial_adapter_openai.utils.image import ImageDataURL, ImageMetadata
+from aidial_adapter_openai.dial_api.storage import FileStorage
+from aidial_adapter_openai.utils.image import ImageMetadata
 from aidial_adapter_openai.utils.log_config import logger
 from aidial_adapter_openai.utils.multi_modal_message import (
     MultiModalMessage,
     create_image_content_part,
     create_text_content_part,
 )
-from aidial_adapter_openai.utils.storage import FileStorage
-
-
-class TransformationError(BaseModel):
-    image_fails: List[ImageFail]
-
-
-async def transform_message(
-    file_storage: Optional[FileStorage],
-    message: dict,
-) -> MultiModalMessage | TransformationError:
-    content = message.get("content", "")
-    custom_content = message.get("custom_content", {})
-    attachments = custom_content.get("attachments", [])
-    logger.debug(f"original attachments: {attachments}")
-
-    message = {k: v for k, v in message.items() if k != "custom_content"}
+from aidial_adapter_openai.utils.resource import Resource
+from aidial_adapter_openai.utils.text import decapitalize
+
+# Officially supported image types by GPT-4 Vision, GPT-4o
+SUPPORTED_IMAGE_TYPES = ["image/jpeg", "image/png", "image/webp", "image/gif"]
+SUPPORTED_FILE_EXTS = ["jpg", "jpeg", "png", "webp", "gif"]
+
+
+@dataclass(order=True, frozen=True)
```

Comment:
Why don't just use pydantic `BaseModel` here?
```python
class TransformationError(BaseModel):
    name: str
    message: str

    class Config:
        frozen = True
```

Reply:
Because I want the error to be printed to the user in a **lexicographical** order.

`@dataclass(order=True)` automatically created the required comparison methods (such as `__le__`).

Whereas the `Config.frozen = True` leads to the generation of `__hash__`, which is required for storing an object in a set, but it is not very good if you want to display information to the user. 

It doesn't define any particular order at all. The hash of the same string may change from one program run to another, and how the hash is computed is an implementation detail.